packages/viz/src/components/ConfusionMatrix/utils.ts

```diff
@@ -0,0 +1,189 @@
+
+import { HvConfusionMatrixFormat, HvConfusionMatrixValuesProps } from "./types";
+
+export const useColorScale = ({
+  data,
+  hasDelta,
+  hasSemanticScale,
+  customColorScale,
+  filterKey,
+}: {
+  data: ColumnTable;
+  hasDelta: boolean;
+  hasSemanticScale: boolean;
+  filterKey: string;
+  customColorScale?: [string, string];
+}) => {
+  const { activeTheme, selectedMode } = useTheme();
+
+  const colorScale = useMemo(() => {
+    if (hasSemanticScale || (hasDelta && customColorScale == null)) {
+      return;
+    }
+
+    const flatData = data
+      .columnNames()
+      .filter((p) => p !== filterKey)
+      .reduce((acc: number[], c: string) => {
+        return [...acc, ...data.array(c)];
+      }, []);
+    const max = Math.max(...flatData);
+    const min = Math.min(...flatData);
+
+    return {
+      color: customColorScale || [
+        activeTheme?.colors.modes[selectedMode].base_light || "",
+        activeTheme?.colors.modes[selectedMode].cat3 || "",
+      ],
+      max,
+      min,
+    };
+  }, [
+    activeTheme?.colors.modes,
+    customColorScale,
+    data,
+    filterKey,
+    hasDelta,
+    hasSemanticScale,
+    selectedMode,
+  ]);
+
+  return colorScale;
+};
+
+export const useSeries = ({
+  data,
+  filterKey,
+  hasDeltaScale,
+  valuesProps,
+}: {
+  data: ColumnTable;
+  filterKey: string;
+  hasDeltaScale: boolean;
+  valuesProps?: HvConfusionMatrixValuesProps;
+}) => {
+  const { activeTheme, selectedMode } = useTheme();
+
+  const getDeltaColor = useCallback(
+    (value: number, diagonal: boolean) => {
+      if ((diagonal && value > 0) || (!diagonal && value < 0)) {
+        return activeTheme?.colors.modes[selectedMode].positive;
+      }
+      if ((diagonal && value < 0) || (!diagonal && value > 0)) {
+        return activeTheme?.colors.modes[selectedMode].negative;
+      }
+
+      return activeTheme?.colors.modes[selectedMode].base_light;
+    },
+    [activeTheme?.colors.modes, selectedMode]
+  );
+
+  const chartSeries = useMemo(() => {
+    return {
+      series: {
+        id: `series~${filterKey}`,
+        type: "heatmap",
+        label: { show: true, ...valuesProps },
+        emphasis: {
+          disabled: true,
+        },
+        data: data
+          .columnNames()
+          .filter((p) => p !== filterKey)
+          .reduce((acc: (string | number)[][], c: string, j) => {
+            const row: (string | number)[][] = data
+              .array(c)
+              .reduce((racc, rv, i) => {
+                return [
+                  ...racc,
+                  {
+                    value: [data.array(filterKey)[i], c, rv != null ? rv : "-"],
+                    ...(hasDeltaScale && {
+                      visualMap: false,
+                      itemStyle: {
+                        color: getDeltaColor(rv, i === j),
+                      },
+                    }),
+                  },
+                ];
+              }, []);
```

Comment:
I worry about the performance in this block, since `data` can grow big and this is quadratic.

Can we use the `array.push(newEl)` instead of `[...array, newEl]`? See: https://jpcamara.com/2023/03/07/making-tanstack-table.html#using-push-directly